File: app/src/main/java/com/androplay/airplay/AirPlayService.java
package com.androplay.airplay;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.androplay.R;

/**
 * AirPlay Service - Background service for AirPlay functionality
 * Runs as a foreground service to maintain connection and handle AirPlay requests
 */
public class AirPlayService extends Service {
    
    private static final String TAG = "AirPlayService";
    private static final String CHANNEL_ID = "AirPlayServiceChannel";
    private static final int NOTIFICATION_ID = 1;
    
    private final IBinder binder = new AirPlayBinder();
    private boolean isRunning = false;
    
    public class AirPlayBinder extends Binder {
        public AirPlayService getService() {
            return AirPlayService.this;
        }
    }
    
    @Override
    public void onCreate() {
        super.onCreate();
        Log.i(TAG, "AirPlayService created");
        createNotificationChannel();
    }
    
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.i(TAG, "AirPlayService started");
        
        // Start as foreground service
        startForeground(NOTIFICATION_ID, createNotification());
        
        // Initialize AirPlay server
        initializeAirPlayServer();
        
        isRunning = true;
        
        return START_STICKY;
    }
    
    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.i(TAG, "AirPlayService destroyed");
        stopAirPlayServer();
        isRunning = false;
    }
    
    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "AirPlay Service",
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("AirPlay service for AndroPlay");
            
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }
    
    private Notification createNotification() {
        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("AndroPlay AirPlay")
                .setContentText("AirPlay service is running")
                .setSmallIcon(R.drawable.ic_airplay)
                .setPriority(NotificationCompat.PRIORITY_LOW)
                .build();
    }
    
    private void initializeAirPlayServer() {
        // TODO: Implement AirPlay server initialization
        // This would typically involve:
        // 1. Setting up network sockets
        // 2. Implementing AirPlay protocol handlers
        // 3. Starting device discovery
        // 4. Setting up media streaming endpoints
        
        Log.i(TAG, "Initializing AirPlay server...");
        
        // Placeholder for actual implementation
        new Thread(() -> {
            try {
                // Simulate server initialization
                Thread.sleep(2000);
                Log.i(TAG, "AirPlay server initialized successfully");
                
                // Start listening for AirPlay connections
                startListening();
                
            } catch (InterruptedException e) {
                Log.e(TAG, "AirPlay server initialization interrupted", e);
            }
        }).start();
    }
    
    private void startListening() {
        // TODO: Implement actual AirPlay listening
        Log.i(TAG, "Starting to listen for AirPlay connections...");
        
        // This would typically:
        // 1. Bind to network interfaces
        // 2. Listen on AirPlay ports (7000, 5353 for mDNS)
        // 3. Handle incoming connections
        // 4. Process AirPlay requests
    }
    
    private void stopAirPlayServer() {
        Log.i(TAG, "Stopping AirPlay server...");
        // TODO: Implement server cleanup
    }
    
    public boolean isServiceRunning() {
        return isRunning;
    }
}

File: app/src/main/java/com/androplay/service/AirPlayService.kt
package com.androplay.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.os.Build
import android.os.IBinder
import com.androplay.MainActivity

class AirPlayService : Service() {

    companion object {
        const val ACTION_START = "com.androplay.action.START_RECEIVER"
        const val ACTION_STOP = "com.androplay.action.STOP_RECEIVER"
        const val CHANNEL_ID = "androplay_channel"
        const val NOTIFICATION_ID = 1
    }

    private val manager by lazy { AirPlayManager.getInstance(this) }
    private var running = false
    private val stateCallback: (AirPlayConnectionState, StreamInfo, String?) -> Unit = { state, _, _ ->
        if (running) {
            when (state) {
                AirPlayConnectionState.AdvertisingOnly -> getSystemService(NotificationManager::class.java)
                    .notify(NOTIFICATION_ID, buildNotification("Discoverable only — streaming unavailable"))
                AirPlayConnectionState.Error -> {
                    running = false
                    stopForeground(STOP_FOREGROUND_REMOVE)
                    stopSelf()
                }
                else -> Unit
            }
        }
    }

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        manager.registerStateCallback(stateCallback)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action == ACTION_STOP) {
            if (running) manager.stop()
            running = false
            stopSelf()
            return START_NOT_STICKY
        }
        if (running) return START_NOT_STICKY

        startForeground(NOTIFICATION_ID, buildNotification("Starting receiver"))
        val settings = ReceiverSettingsStore(this).load()
        if (!manager.start(settings)) {
            stopForeground(STOP_FOREGROUND_REMOVE)
            stopSelf(startId)
            return START_NOT_STICKY
        }
        running = true
        getSystemService(NotificationManager::class.java)
            .notify(NOTIFICATION_ID, buildNotification(
                if (manager.isDiscoveryOnly) "Publishing AirPlay discovery" else "Waiting for AirPlay connection"
            ))
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onDestroy() {
        if (running) manager.stop()
        manager.unregisterStateCallback(stateCallback)
        super.onDestroy()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "AndroPlay Service",
                NotificationManager.IMPORTANCE_LOW
            )
            val manager = getSystemService(NotificationManager::class.java)
            manager.createNotificationChannel(channel)
        }
    }

    private fun buildNotification(message: String): Notification {
        val pendingIntent = PendingIntent.getActivity(
            this, 0,
            Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_IMMUTABLE
        )

        return Notification.Builder(this, CHANNEL_ID)
            .setContentTitle("AndroPlay")
            .setContentText(message)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .build()
    }
}
